VideoPress: only enqueue the bridge script when the player is rendered

jgcaruso opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes #24765

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Moved front end bridge script enqueueing to VideoPress player render functions
* removed duplicate enqueue statement for admin pages (different version number was causing double enqueue)

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
Nope
#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Use a site with VideoPress enabled
* Create a post with a VideoPress shortcode
* Create a post with a `/video` block
* Create a post and paste a VideoPress url in the page (uses oembed)
* Create a post with no VideoPress content in it
* View the pages in the front end, the `videopress-token-bridge.js` script should be rendered only on pages with VideoPress content
* View the homepage, it should also not load the bridge script unless your post summaries include the VideoPress player
* Make one or more of the videos on your posts private
* View the post on the front end to ensure the bridge is working correctly
* Visit `/wp-admin`, private videos in the post editor and the Media Library should still be viewable

jgcaruso committed:
[not verified] fixed require statement

jgcaruso committed:
[not verified] fixed require statement

jgcaruso committed:
[not verified] changelog

jgcaruso committed:
[not verified] replaces string index check with regex

jgcaruso committed:
[not verified] remove duplicate token bridge enqueue statement

jgcaruso committed:
[not verified] call enqueue functions directly as it may be too late to register events.

roundhill reviewed (commented):
Testing really well, just needs conflicts fixed for ✅

roundhill reviewed (dismissed):
👍

sdixon194 commented:
Looks like it doesn't like the unused `$attr` attribute in `video_enqueue_bridge_when_oembed_present`. If it's not needed, can it be removed, or otherwise throw a `phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable` on there?

jgcaruso commented:
> Looks like it doesn't like the unused `$attr` attribute in `video_enqueue_bridge_when_oembed_present`. If it's not needed, can it be removed, or otherwise throw a `phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable` on there?

I think I'll put the ignore on there... those are part of the filter's signature so I'd rather not mess with it, unless that is a normal thing to do?

jgcaruso committed:
[not verified] phpcs ignore unused params in filter handler

sdixon194 commented:
> those are part of the filter's signature so I'd rather not mess with it, unless that is a normal thing to do?

Yup, that makes sense and works for me!

jeherve reviewed (dismissed):
Looks like the tests are failing now:

```
PHP Fatal error:  Class 'VideoPress_Scheduler' not found in /tmp/wordpress-latest/src/wp-content/plugins/jetpack/modules/videopress/class.jetpack-videopress.php on line 71
  ..............................
  Fatal error: Class 'VideoPress_Scheduler' not found in /tmp/wordpress-latest/src/wp-content/plugins/jetpack/modules/videopress/class.jetpack-videopress.php on line 71
```

jgcaruso committed:
use new token bridge class

jgcaruso committed:
remove general enqueue hook for non-admin pages

jgcaruso committed:
copy changelog to new location

jgcaruso committed:
whitespace

jgcaruso commented:
Rebased and updated the files since a refactor in another PR moved around a bunch of the files.

Also noticed an issue with Private videos no longer loading as a result of that refactor, which is fixed separately in https://github.com/Automattic/jetpack/pull/25638

jgcaruso committed:
fixed videopress changelog type

jgcaruso committed:
removed no longer necessary import

leogermani commented:
@jgcaruso One question related to this. I was under the impression that the class VideoPress_Player was only used by the shortcode. Is that true? Or is it only used by the block in some way?

Also, I thought the jwt token bridge was used by all player instances, either block or shortcodes. Is it?

For context, we are cherry picking what will be migrated to the package (to be shared with the stand-alone plugin) and what will not. See pe4Cmx-40-p2

jgcaruso commented:
> I was under the impression that the class VideoPress_Player was only used by the shortcode. Is that true? Or is it only used by the block in some way?

The code for the VideoPress_Player does appear to only be used by the shortcode. I haven't noticed any other references to it. Now that you call this out, I'm wondering why I put the `embed_oembed_html` filter handler in the `shortcode.php` file. It might make more sense somewhere else.

Given you're reorganizing all of this code, where would you advise me to put the `embed_oembed_html` filter handler so that it gets included where it needs to for both the standalone plugin and Jetpack? This is the handler that will enqueue the jwt token bridge for the block when present on a page.

> Also, I thought the jwt token bridge was used by all player instances, either block or shortcodes. Is it?

The jwt token bridge is used by all player instances. More specifically, it is used directly by the videopress iframe that gets rendered by them in order for private videos to load correctly.

leogermani commented:
> I'm wondering why I put the embed_oembed_html filter handler in the shortcode.php file. It might make more sense somewhere else.

The custom oembed regexp registration used to live there and maybe you had this on your memory.

This has been already moved to the package:

https://github.com/Automattic/jetpack/blob/trunk/projects/packages/videopress/src/class-initializer.php#L72

So I guess this filter should also be added there in the initializer.

> The jwt token bridge is used by all player instances. More specifically, it is used directly by the videopress iframe that gets rendered by them in order for private videos to load correctly.

Gotcha, and it's added via the oembed filter. So it definitely needs to go in the package.

jgcaruso committed:
move oembed filter to initializer class

jgcaruso commented:
@leogermani thanks for that! I've moved the oembed filter in with the oembed initializers so it should be good to go now.

jgcaruso committed:
Merge branch 'trunk' into update/videopress-only-enqueue-bridge-when-player-is-rendered

leogermani commented on projects/packages/videopress/src/class-initializer.php:
```diff
@@ -76,5 +76,24 @@ public static function register_oembed_providers() {
 		// This is needed as it's not supported in oEmbed discovery
 		wp_oembed_add_provider( '|^https?://v\.wordpress\.com/([a-zA-Z\d]{8})(.+)?$|i', 'https://public-api.wordpress.com/oembed/?for=' . $host, true ); // phpcs:ignore WordPress.WP.CapitalPDangit.Misspelled
+
+		add_filter( 'embed_oembed_html', array( 'Automattic\Jetpack\VideoPress\Initializer', 'video_enqueue_bridge_when_oembed_present' ), 10, 4 );
```
nitpick: can we use `__CLASS__` here?

leogermani reviewed (dismissed):
Tests well!

Left a minor comment

leogermani commented on projects/packages/videopress/src/class-initializer.php:
```diff
@@ -76,5 +76,24 @@ public static function register_oembed_providers() {
+	 * @return string|false
+	 */
+	public static function video_enqueue_bridge_when_oembed_present( $cache, $url, $attr, $post_ID ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		if ( preg_match( '/https?:\/\/(video.wordpress.com|videopress.com)\/(v|embed)\//', $url ) ) {
```
Don`t we need to add support to `v.wordpress.com` as well?

jgcaruso replied:
I don't think I've ever seen `v.wordpress.com` actually used anywhere as an embed, and that url doesn't seem to work for viewing videos on its own. Its possible it was an old way of accessing videos.

I suppose I could add it for completeness.

jgcaruso replied:
I added a 2nd condition for the `https://v.wordpress.com/GUID` embeds. Since the regex just translates them into a standard embed it was easy to test as well, even though the url itself doesn't do anything.

jgcaruso committed:
cleaner reference for current class

jgcaruso committed:
add support for legacy v.wordpress.com/GUID urls

jgcaruso committed:
Merge branch 'trunk' into update/videopress-only-enqueue-bridge-when-player-is-rendered

jeherve reviewed (commented):
I believe this will need one last update with trunk before we can merge it.

jeherve commented on projects/plugins/jetpack/changelog/update-videopress-only-enqueue-bridge-when-player-is-rendered:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: other
+
+Only add the VideoPress bridge script when a VideoPress player will be rendered on the page.
```
I would suggest making this a bit more end-user friendly. Something like:

```suggestion
VideoPress: ensure that the player's resources are only loaded when necessary, when a video was added to the page.
```

jeherve commented on projects/packages/videopress/src/class-initializer.php:
```diff
@@ -76,5 +76,25 @@ public static function register_oembed_providers() {
+	 * @return string|false
+	 */
+	public static function video_enqueue_bridge_when_oembed_present( $cache, $url, $attr, $post_ID ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		if ( preg_match( '/https?:\/\/(video.wordpress.com|videopress.com)\/(v|embed)\//', $url )
```
Should we be a bit more explicit, like below, to avoid matching for things like `https://mysite.com/https://videopressscom/v/AxWYhwqp`:

```suggestion
		if ( preg_match( '/^https?:\/\/(video\.wordpress\.com|videopress\.com)\/(v|embed)\//', $url )
```

Maybe you could add a test for this, too, to ensure we catch the right things?

jgcaruso replied:
Thanks for catching that... I just used the existing regex from earlier in the file and didn't consider that it could have been incorrect.

jgcaruso committed:
Update projects/plugins/jetpack/changelog/update-videopress-only-enqueue-bridge-when-player-is-rendered

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

jgcaruso committed:
Update projects/packages/videopress/src/class-initializer.php

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

jgcaruso committed:
Merge branch 'trunk' into update/videopress-only-enqueue-bridge-when-player-is-rendered

jgcaruso committed:
linter

jgcaruso committed:
add test for enqueue function for bridge script

jgcaruso committed:
add test cases for `embed` urls as well

jgcaruso commented:
@jeherve I've made the changes and added a test for the function that ensures the proper video url types trigger the enqueue function.

There are a couple of github checks that are failing that I'm not sure how to address, unless they're ok?

One is the `Linting / Lock files are up to date (pull_request)` and the other is a file sync check, but it might be easier to clean that one up after the fact since that code has changed a bit and should be cleaner once this PR goes in. Based on whats missing (looks like new code if I'm reading the file sync checker tool correctly for `projects/plugins/jetpack/modules/videopress/class.videopress-gutenberg.php`), I suspect it was skipped previously as well.

jeherve reviewed (changes requested):
Thanks for adding the tests!

> One is the Linting / Lock files are up to date (pull_request) and the other is a file sync check, but it might be easier to clean that one up after the fact since that code has changed a bit and should be cleaner once this PR goes in. Based on whats missing (looks like new code if I'm reading the file sync checker tool correctly for projects/plugins/jetpack/modules/videopress/class.videopress-gutenberg.php), I suspect it was skipped previously as well.

You should be able to fix the linting / lock file issues by updating with latest trunk once again, and running the commands suggested by the failure messages in this view:
https://github.com/Automattic/jetpack/pull/24985/files
(since you made changes to a composer.json file, you have to update the matching composer.lock files, as well as lock files for projects that have the VideoPress package as a dependency).

The new test seems to be failing though:
```PHP Fatal error:  Declaration of Automattic\Jetpack\InitializerTest::setUp() must be compatible with PHPUnit\Framework\TestCase::setUp(): void in /home/runner/work/jetpack/jetpack/projects/packages/videopress/tests/php/test-class-initializer.php on line 18
  
  Fatal error: Declaration of Automattic\Jetpack\InitializerTest::setUp() must be compatible with PHPUnit\Framework\TestCase::setUp(): void in /home/runner/work/jetpack/jetpack/projects/packages/videopress/tests/php/test-class-initializer.php on line 18
```

jgcaruso committed:
Merge branch 'trunk' into update/videopress-only-enqueue-bridge-when-player-is-rendered

jgcaruso committed:
ran fixup-project-versions.sh

jgcaruso committed:
fix setUp/tearDown signatures

jgcaruso committed:
fix test setup/teardown to be compatible with all versions of php

jgcaruso committed:
Merge branch 'trunk' into update/videopress-only-enqueue-bridge-when-player-is-rendered

jgcaruso committed:
fixup project versions again after conflict from merge

jgcaruso committed:
Merge branch 'trunk' into update/videopress-only-enqueue-bridge-when-player-is-rendered

jgcaruso commented:
@jeherve Ok I was able to fix up the composer files using the `fixup` tool and tests are working again after I looked at some of the other tests to see how they were getting around the php 5.6 lack of support for `:void`.

The file sync issue also looks like its gone after the deploy of a diff that was on hold for a while.

This PR **_should_** be good to go now 🤞

jgcaruso committed:
Merge branch 'trunk' into update/videopress-only-enqueue-bridge-when-player-is-rendered

jgcaruso commented:
deployed r250884-wpcom
